docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+- **Extensibility/Transformation**: Introduce mechanism for allowing Cluster specific transformations of a ClusterClass (e.g. inject API Endpoint for CAPV, customize machine image by version etc.)
+- **Adoption**: Providing a way to convert existing clusters into managed cluster topologies.
+- **Observability**: Build an SDK and enhance the Cluster object status to surface a summary of the status of managed topology.
+- **Healthchecks**: Augment ClusterClass to include MachineHealthCheck refs to manage the health of the managed topology.
+- **Lifecycle integrations**: Extend ClusterClass to include lifecycle management integrations such as MachineHealthCheck and Cluster Autoscaler to manage the state and health of the managed clusters.
+
+However we are intentionally leaving them out from this initial iteration for the following reasons:
+- We want the community to reach a consensus on cornerstone elements of the design before iterating on additional features.
+- We want to enable starting the implementation of the required scaffolding and the initial support for managed topologies as soon as possible, so we can surface problems which are not easy to identify at this stage of the proposal.
+- We would like the community to rally in defining use cases for the advanced features, help in prioritizing them, so we can chart a more effective roadmap for the next steps.
+
+
+## Proposal
+
+This proposal enhances the `Cluster` object to create managed cluster topologies using the `ClusterClass` object.
+
+[PlantUML](http://plantuml.com) is the preferred tool to generate diagrams,
+place your `.plantuml` files under `images/` and run `make diagrams` from the docs folder.
+
+### User Stories
+
+#### Story 1 - Use ClusterClass to easily stamp clusters
+As an end user, I want to use one `ClusterClass` to create multiple managed cluster topologies of similar flavor.
+- Rather than recreating the KCP and MD objects for every cluster that needs to be provisioned, the end user can create a template once and reuse it to create multiple clusters with similar configurations.
+
+#### Story 2 - Easier UX for kubernetes version upgrades
+For an end user, the UX to update the kubernetes version of the control plane and worker nodes in the cluster should be easy.
+- Instead of individually modifying the KCP and each MachineDeployment, updating a single option should result in k8s version updates for all the CP and worker nodes.
+
+**Note**: In order to complete the user story for all the providers, some of the advanced features (such as Extensibility/Transformation) are required. However, getting this in place even only for a subset of providers allows us to build and test a big chunk of the entire machinery.
+
+#### Story 3 - Easier UX for scaling workers nodes
+As an end user, I want to be able to easily scale up/down the number of replicas for each set of worker nodes in the cluster.
```

Comment:
Can we elaborate on the real use case for this UX? e.g is this thinking on having a single point fo scaling machines evenly across AZs? Is it something else?

Reply:
Instead of having to modify one or more MachineDeployments, you can simply change the desired worker count of the Cluster resource. If they're managed that's also the only way, as you probably shouldn't touch the resources created by the controller.

Reply:
>Instead of having to modify one or more MachineDeployments, you can simply change the desired worker count of the Cluster resource. If they're managed that's also the only way, as you probably shouldn't touch the resources created by the controller.

Make sense so It'd be nice to illustrate a bit more here user story/scenarios/topologies where this UX and functionality (ability to scale multiple scalable resources in a single CR) will come in handy rather than describing the functionality itself. E.g for the opposite use case: As a cluster operator I want to provide certain users (e.g Autoscaler, Tester...) with granular perms to scale machines but prevent them from modifying cluster properties so I want this ability to be scoped in different CRs.

>As I'm not familiar how CAPI handles AZs this is just a guess, but since the managed workers are also created using a MachineDeployment, which is created from some form of template, I guess it will continue to depend on the provider in use how distributing nodes across AZs has to be configured.

Nothing does that for you today, each pool of nodes for a machineDeployment belong to a single AZ unless you are using MachinePools. The functionality described here might ease that manual task for an operator, E.g As a cluster operator I want to have a single point to scale my MachineDeployments so I can easily ensure they result in an even number of workers across AZs to be more tolerant to failures.